File: Java/Sorting/BeadSort.java
import java.util.Arrays;
import java.util.Scanner;

public class BeadSort {
    public static void display(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void beadSort(int[] arr) {
        int max = Arrays.stream(arr).max().getAsInt();
        int len = arr.length;
        int[][] beads = new int[len][max];

        for (int i = 0; i < len; i++) {
            for (int j = 0; j < arr[i]; j++) {
                beads[i][j] = 1;
            }
        }

        for (int j = 0; j < max; j++) {
            int sum = 0;
            for (int i = 0; i < len; i++) {
                sum += beads[i][j];
                beads[i][j] = 0;
            }
            for (int i = len - sum; i < len; i++) {
                beads[i][j] = 1;
            }
        }

        for (int i = 0; i < len; i++) {
            int j;
            for (j = 0; j < max && beads[i][j] != 0; j++) {
            }
            arr[i] = j;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the size of the array: ");
        int n = scanner.nextInt();

        System.out.print("Enter the elements of the array: ");
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }

        System.out.print("Original array: ");
        display(arr);

        beadSort(arr);

        System.out.print("Sorted array: ");
        display(arr);
    }
}

File: C/Sorting/Bead_sort.c
#include <stdio.h>
#include <stdlib.h>

void display(const int *arr, int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

void beadSort(int *arr, size_t len) {
    int i, j, max, sum;
    unsigned char *beads;

    for (i = 1, max = arr[0]; i < len; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }

    beads = calloc(1, max * len);

    for (i = 0; i < len; i++) {
        for (j = 0; j < arr[i]; j++) {
            beads[i * max + j] = 1;
        }
    }

    for (j = 0; j < max; j++) {
        for (sum = i = 0; i < len; i++) {
            sum += beads[i * max + j];
            beads[i * max + j] = 0;
        }
        for (i = len - sum; i < len; i++) {
            beads[i * max + j] = 1;
        }
    }

    for (i = 0; i < len; i++) {
        for (j = 0; j < max && beads[i * max + j]; j++) {
        }
        arr[i] = j;
    }
    free(beads);
}

int main() {
    int n;
    printf("Enter the size of the array: ");
    scanf("%d", &n);

    int *arr = (int *)malloc(n * sizeof(int));
    printf("Enter the elements of the array: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    printf("Original array: ");
    display(arr, n);

    beadSort(arr, n);

    printf("Sorted array: ");
    display(arr, n);

    free(arr);
    return 0;
}
